src/cp_als.jl

```diff
@@ -0,0 +1,80 @@
+export cp_als
+
+## The Alternating Least-Squares (ALS) algorithm for CP-decomposition with ITensors.jl
+# Based on "Tensor Decompositions and Applications", Tamara G. Kolda and Brett W. Bader. SIAM Review 51(3), 2009
+# Porting the code from https://www.kaggle.com/nicw102168/rank-of-random-2x2x2-tensors
+
+using ITensors
+using ITensors.LinearAlgebra
+
+
+function cp_als(X, R; maxiter=1000, test_period=1, tol=1e-10)
+    # initialize with spheric-random colunms
+    A = map([randomITensor(Iₙ, Index(R)) for Iₙ in inds(X)]) do An
+        aλ = columnnorms(An)
+        rr = Index(R)
+        iλ = diagITensor(aλ.^-1, rr, inds(An)[2])
+        replaceind(An, inds(An)[2], rr) * iλ
+    end
+    cp_als_(X, R, A, maxiter=maxiter, test_period=test_period, tol=tol)
+end
+
+function cp_als_(X, R, A; maxiter=1000, test_period=1, tol=1e-10)
+    N = order(X)
+
+    rind = [inds(An)[2] for An in A]
+    λ = nothing
+    iteration = nothing
+
+    for it in 1:maxiter
+        iteration=it
+
+        for n in 1:N
+            V = reduce((a,b)->a.*b, array(A[m])' * array(A[m]) for m in 1:N if m != n) ## is there a better, ITensor way to do that?
+            Vinv = ITensors.LinearAlgebra.pinv(V)
+
+            W = reduce(khatrirao, [A[m] for m in N:-1:1 if m != n])
+            Xn = unfold(X,n)
+            qq = diagITensor(ones(inds(Xn)[2].space), inds(Xn)[2], inds(W)[1])
+
+            rr = Index(R)
+            newAn = Xn * qq * W * itensor(Vinv, inds(W)[2], rr)
```

Comment:
Could you explain what this operation is doing?

Reply:
Here we compute the second row in the loop of the Figure 3.3 algorithm. The `W` is the central part in the expression where we compute the Khatri-Rao products across the matrices, skipping the current dimenson, similar to before. `Vinv` and Xn are the other clear elements in that expression. The rest I'd have to think more to remember, it's just what I figured out was a way to get all of this together... I suppose it must be possible to write simply as `Xn * W * Vinv`, but I messed up the tensor shapes somehow.